data/scripts.js

```diff
@@ -296,158 +259,252 @@ let BattleScripts = {
-		if (!hitResult) {
-			if (hitResult !== null) {
+		let moveResult = !!targets.length;
+		if (!moveResult && !atLeastOneFailure) pokemon.moveThisTurnResult = null;
+		if (move.spreadHit) this.attrLastMove('[spread] ' + targets.join(','));
+		return moveResult;
+	},
+	tryImmunityEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryImmunity', targets, pokemon, move);
+		for (let i = 0; i < targets.length; i++) {
+			if (hitResults[i] === false) {
 				if (!move.spreadHit) this.attrLastMove('[miss]');
-				this.add('-miss', pokemon, target);
+				this.add('-miss', pokemon, targets[i]);
+			} else {
+				hitResults[i] = hitResults[i] || false;
 			}
-			return false;
 		}
-
+		return hitResults;
+	},
+	typeImmunity(targets, pokemon, move) {
 		if (move.ignoreImmunity === undefined) {
 			move.ignoreImmunity = (move.category === 'Status');
 		}
 
-		if (this.gen < 7 && (!move.ignoreImmunity || (move.ignoreImmunity !== true && !move.ignoreImmunity[move.type])) && !target.runImmunity(move.type, true)) {
-			return false;
+		let hitResults = [];
+		for (let i = 0; i < targets.length; i++) {
+			hitResults[i] = (move.ignoreImmunity && (move.ignoreImmunity === true || move.ignoreImmunity[move.type])) || targets[i].runImmunity(move.type, true);
 		}
 
-		hitResult = this.runEvent('TryHit', target, pokemon, move);
-		if (!hitResult) {
-			if (hitResult === false) {
-				this.add('-fail', pokemon);
-				this.attrLastMove('[still]');
-			} else if (hitResult === this.NOT_FAILURE) {
-				return hitResult;
-			}
-			return false;
+		return hitResults;
+	},
+	tryHitEvent(targets, pokemon, move) {
+		let hitResults = this.runEvent('TryHit', targets, pokemon, move);
+		if (!hitResults.includes(true) && hitResults.includes(false)) {
+			this.add('-fail', pokemon);
+			this.attrLastMove('[still]');
 		}
-
-		if (this.gen >= 7 && (!move.ignoreImmunity || (move.ignoreImmunity !== true && !move.ignoreImmunity[move.type])) && !target.runImmunity(move.type, true)) {
-			return false;
+		for (let i = 0; i < targets.length; i++) {
+			if (hitResults[i] !== this.NOT_FAILURE) hitResults[i] = hitResults[i] || false;
 		}
-		if (move.flags['powder'] && target !== pokemon && !this.getImmunity('powder', target)) {
-			this.debug('natural powder immunity');
-			this.add('-immune', target);
-			return false;
+		return hitResults;
+	},
+	powderImmunity(targets, pokemon, move) {
+		let hitResults = [];
+		if (!move.flags['powder']) {
+			for (let i = 0; i < targets.length; i++) {
+				hitResults[i] = true;
+			}
+			return hitResults;
 		}
-		if (this.gen >= 7 && move.pranksterBoosted && pokemon.hasAbility('prankster') && target.side !== pokemon.side && !this.getImmunity('prankster', target)) {
-			this.debug('natural prankster immunity');
-			if (!target.illusion) this.hint("In Gen 7, Dark is immune to Prankster moves.");
-			this.add('-immune', target);
-			return false;
+		for (let [i, target] of targets.entries()) {
```

Comment:
Oh hey, here we're using `entries` :). Can you `const` with `entries` (everywhere in this file)?